networkit/cpp/centrality/PageRank.cpp

```diff
@@ -71,17 +71,39 @@ void PageRank::run() {
 
     handler.assureRunning();
 
-    const auto sum = G.parallelSumForNodes([&](const node u) { return scoreData[u]; });
+    if (!normalized) {
+        const auto sum = G.parallelSumForNodes([&](const node u) { return scoreData[u]; });
+
+        // make sure scoreData sums up to 1
+        assert(!Aux::NumericTools::equal(sum, 0.0, 1e-15));
+        G.parallelForNodes([&](const node u) { scoreData[u] /= sum; });
+    } else {
+        // calculate sum of the Pagerank of dangling Nodes for normalization
+        const auto sum = G.parallelSumForNodes([&](const node u) {
+            if (G.degree(u) == 0.0)
+                return scoreData[u];
+            return 0.0;
+        });
+
+        const auto normFactor = (1.0 / n) * (tol + ((1.0 - tol) * sum));
 
-    // make sure scoreData sums up to 1
-    assert(!Aux::NumericTools::equal(sum, 0.0, 1e-15));
-    G.parallelForNodes([&](const node u) { scoreData[u] /= sum; });
+        G.parallelForNodes([&](const node u) { scoreData[u] /= normFactor; });
+    }
 
     hasRun = true;
 }
 
 double PageRank::maximum() {
-    return 1.0; // upper bound, could be tighter by assuming e.g. a star graph with n nodes
+    if (!normalized)
+        return 1.0; // upper bound, could be tighter by assuming e.g. a star graph with n nodes
+
+    auto max = scoreData[0]; // Unlike regular Page Rank there isn't a universal upper bound for
+                             // normalized page rank. So we need to iterate.
+    G.balancedParallelForNodes([&](const node u) {
+        if (max < scoreData[u])
+            max = scoreData[u];
```

Comment:
This is not a true maximum calculation due to race conditions, i.e., if there are two threads that find a larger value at the same time, there is no guarantee that the larger of the two will be stored. There is [Aux::Parallel::atomic_max()](https://github.com/networkit/networkit/blob/master/include/networkit/auxiliary/Parallel.hpp) in NetworKit that allows computing the maximum in a thread-safe way. I also think there is no need to use the `balanced` iteration as the computation for all nodes should be identical. Further, I would suggest to move this calculation into the run method and store the maximum in a member variable.